ucb-rit/coldfront
# Add support for allocation renewal requests for FCAs for the current allocation year

**Changes**
- Features
    - Added a multi-part form to allow certain users to renew a particular PI's allocation and update their pooling preference for Faculty Compute Allowances (FCAs) for the current allocation year (06/01 - 05/31).
        - The allocation of PI `A` may be renewed by any manager or PI of any `Project` of which PI `A` has role `Principal Investigator`.
        - This is accessible from the Project List view, via a "Renew a PI's Allowance" button.
        - Each PI may have at most one FCA per allocation year, so any that have pending/processing/completed `AllocationRenewalRequests` or pending/processing/completed `SavioProjectAllocationRequests` are not selectable.
            - The same is true for requesting a new/pooled `Project`.
    - Added a shorter form to renew a particular PI under a particular `Project`, which is the most common case.
        - This is accessible to the managers and PIs of a Project, from the Project Detail view, via a "Renew Allowance" button.
    - Added an admin view for approving/denying renewal requests, under "Project Renewal Requests".
        - For FCAs, there are two options for approval steps:
            - If a new Project was requested, the corresponding `SavioProjectAllocationRequest` must be processed first.
            - Otherwise, the admin should verify that the PI is still eligible for an allowance.
        - Staff users have read-only access.
- Database Changes
    - Added the following database models:
        - `AllocationPeriod`
        - `AllocationRenewalRequestStatusChoice`
        - `AllocationRenewalRequest`
    - Added a new `ProjectStatusChoice` named `Inactive` for FCAs that were not renewed.
- Management Commands
    - Added a command for creating a list of pre-defined `AllocationPeriod` objects.
    - Added a command for loading existing FCA renewal requests from a JSON file.
    - Added a command for deactivating FCA projects that were not renewed.
    - Modified the `create_staff_group` command to incorporate new permissions relating to these requests, and included it in the Ansible playbook to run on each deployment.
- Testing
    - Added tests for:
        - The utility for processing renewal requests
        - The utility for denying renewal requests
        - Utility methods and signals
        - The behavior of various forms and views

**How to Test**
- Review the code changes and add comments as needed.
- Run preliminary commands.
    - Apply new database migrations.
        ```
        python manage.py migrate
        ```
    - Create the `staff_group` `Group`, or update its permissions, to account for the new type of request.
        ```
        python manage.py create_staff_group
        ```
    - Create `AllocationPeriods`.
        ```
        python manage.py create_allocation_periods
        ```
    - Load existing FCA renewal requests, which came in from Google Forms, from a JSON file, which should be requested separately.
        ```
        python manage.py load_allocation_renewal_requests /path/to/requests.json
        ```
    - Deactivate FCAs that were not renewed.
        ```
        python manage.py deactivate_fca_projects
        ```
- Ensure that the newly-added tests pass.
    ```
    python manage.py test coldfront.core.project.tests
    ```
- Manual Testing
    - Review user-facing language for clarity and consistency with actual behavior.
    - Create some requests at `/project/renew-pi-allocation-landing` and `/project/{pk}/renew`. Note any exceptions.
    - Approve/deny the requests using the new Admin view (Project Renewal Requests). Note any exceptions.

**Notes**
- The currently in-use Google Form for handling FCA renewals should be disabled so that all requests come through the portal.
- Support for renewal requests for the next allocation year, and for different allocation types (e.g., PCAs and ICAs) should be possible by extending this design.
    - There are a number of `TODOs` to this effect.
- The tests added so far are not comprehensive.
    - There are a number of `TODOs` to fill in.
